django/core/handlers/base.py

```diff
@@ -228,44 +175,88 @@ def get_response(self, request):
-        """
-        Pass the exception to the exception middleware. If no middleware
-        return a response for this exception, raise it.
-        """
-        for middleware_method in self._exception_middleware:
-            response = middleware_method(request, exception)
+    def _get_response(self, request):
+        "Returns an HttpResponse object for the given HttpRequest"
+        response = None
+
+        if hasattr(request, 'urlconf'):
+            urlconf = request.urlconf
+            set_urlconf(urlconf)
+            resolver = get_resolver(urlconf)
+        else:
+            resolver = get_resolver()
+
+        resolver_match = resolver.resolve(request.path_info)
+        callback, callback_args, callback_kwargs = resolver_match
+        request.resolver_match = resolver_match
+
+        # Apply view middleware
+        for middleware_method in self._view_middleware:
+            response = middleware_method(request, callback, callback_args, callback_kwargs)
             if response:
                 return response
-        raise
 
-    def handle_uncaught_exception(self, request, resolver, exc_info):
+        wrapped_callback = self.make_view_atomic(callback)
+
+        # TODO: It would be nice if we could move the try/except in _get_response_old,
+        # but that would break semantics.
+        try:
+            response = wrapped_callback(request, *callback_args, **callback_kwargs)
+        except Exception as e:
+            response = self.process_exception_by_middleware(e, request)
+
+        # Complain if the view returned None (a common error).
+        # TODO: New style middlewares will actually be able to handle that, good or not?!
+        if response is None:
+            if isinstance(callback, types.FunctionType):    # FBV
+                view_name = callback.__name__
+            else:                                           # CBV
+                view_name = callback.__class__.__name__ + '.__call__'
+
+            raise ValueError("The view %s.%s didn't return an HttpResponse object. It returned None instead."
+                             % (callback.__module__, view_name))
+
+        # If the response supports deferred rendering, apply template
+        # response middleware and then render the response
+        elif hasattr(response, 'render') and callable(response.render):
+            for middleware_method in self._template_response_middleware:
+                response = middleware_method(request, response)
+                # Complain if the template response middleware returned None (a common error).
+                # TODO: See TODOs above this one.
+                if response is None:
+                    raise ValueError("%s.process_template_response didn't return an "
+                        "HttpResponse object. It returned None instead."
+                        % (middleware_method.__self__.__class__.__name__))
+
+            # TODO: See TODOs above this one.
```

Comment:
Yeah, and here you could again use the hypothetical `self.inner_exception_handler` context manager.

Damn, it'll be nice to clean all this up once old-style middleware is gone!